### ISSUE TITLE
Function Binding Rework

### ISSUE THREAD

**Issue by Mytherin** [cwida/duckdb#263]
Currently we are working on a rework of how functions/function binding works:

* Functions should not have a MatchesArgument function, instead they should have a list of input types with a set of auto cast rules (e.g. TINYINT + SMALLINT can be auto casted to SMALLINT + SMALLINT).
* Functions should support overloading, i.e. there can be multiple functions with the same name but different input parameters (TINYINT + TINYINT, SMALLINT + SMALLINT)
* OperatorExpression should be removed and all current operators should be implemented as functions (e.g. the addition operator "+" will become a set of functions with overloads).
* Adding new functions should be cleaner and require less code than currently

**Comment by Mytherin** [comment]
The current interface that I have in mind for adding functions is as follows:

Simple Function (no overloads)
```c
// prototype:
void AddFunction(string name, vector<SQLType> arguments, SQLType return_type, scalar_function_t function, bool has_side_effects = false, bind_scalar_function_t bind= nullptr, dependency_function_t dep = nullptr);
// usage:
functions.AddFunction("cos", {SQLType(SQLTypeId::DOUBLE)}, SQLType(SQLTypeId::DOUBLE) cos_function);
```

Function with Overloads:
```c
FunctionSet set;
set.AddFunction("add", {SQLType(SQLTypeId::INTEGER)}, SQLType(SQLTypeId:: INTEGER), add_function_integer);
set.AddFunction("add", {SQLType(SQLTypeId::DOUBLE)}, SQLType(SQLTypeId::DOUBLE), add_function_double);
// etc...
```

This would replace the current setup of creating the static classes entirely, and would clean up a lot of headers/etc

**Comment by hawkfish** [comment]
I think mapping prefix/infix operators to functions is the right thing to do. The distinction is purely syntactic and should be handled in the parser. (Besides the term operator is already overladed enough in a database!)

Removing the `MatchesArgument` function seems problematic to me because it makes it hard to specify functions that have multiple incompatible signatures. For example: `IIF` and `MAX` have signatures `type IIF( Boolean, type, type )` and `type MAX( type )`.  This could be worked around if the functions were keyed by _signature_ and resolving a function provided a _list_ of matches that the binder could then choose between. The previous system I worked on did things that way.

This is related to the promotion issue. The implementations deal in physical types, but the binder (not to mention the user!) should only have to worry about logical types like `integer`, `real` and so forth. This suggests that either the binder or the implementation should handle promotion once the logical type matching has been determined. Perhaps function signatures could specify physical types for binding, and the binder could match using logical types and then add casts? Casting could also be deferred to the function implementations, which would make the cast results cache-resident and also allow the function to optimise for particular types instead of just casting them.

As for the amount of code, I think the problem is that SQL effectively requires you to define objects in a 1970s function pointer style. For scalar functions, this is not too onerous because there is usually only one function, but for aggregates it gets complex. Aggregates define an internal, persistent state that requires a number of functions (`initialize`, `update`, `merge`, `finalize`, `deallocate`) as well as the binding information. Windowed aggregates probably add one or two more (I've done this but it was a while ago.) Often there are defaults that can be applied, especially if the aggregate is computed in-place: `initialize` sets to `NULL`, `merge` just aliases to `update`, `finalize` gathers and `deallocate` is a `NOP`, but any notion of "simplifying" should bear this in mind.

Maybe the way to think about it is to imagine the DDL statement. There will likely be a number of arguments, any one of which could be optional and some of which would trigger defaults (e.g., specifying the state as some kind of `struct`?) This sounds like filling in the elements of some kind of record, so maybe the current approach is the right one in the long term. Maybe the complexity is just the current "static function returning function pointer" approach, and we could just simplify it to "struct initialised with function pointers and inheriting some defaults from a base class"?

**Comment by Mytherin** [comment]
> Removing the MatchesArgument function seems problematic to me because it makes it hard to specify functions that have multiple incompatible signatures. For example: IIF and MAX have signatures IIF( Boolean, type, type ): type and MAX( type ): type. This could be worked around if the functions were keyed by signature and resolving a function provided a list of matches that the binder could then choose between. The previous system I worked on did things that way.

For sure, this is exactly what we want to do. We provide a set of function overloads, e.g. for MAX this would be:

MAX(TINYINT) : TINYINT
MAX(SMALLINT) : SMALLINT
MAX(INT) : INT
MAX(BIGINT) : BIGINT
etc...

Then based on the input types we choose one of the overloads, with potential upcasting if required (e.g. if only SIN(DOUBLE) : DOUBLE is defined, the binder adds a cast from INT to DOUBLE before calling the function).

> Maybe the way to think about it is to imagine the DDL statement. There will likely be a number of arguments, any one of which could be optional and some of which would trigger defaults (e.g., specifying the state as some kind of struct?) This sounds like filling in the elements of some kind of record, so maybe the current approach is the right one in the long term. Maybe the complexity is just the current "static function returning function pointer" approach, and we could just simplify it to "struct initialised with function pointers and inheriting some defaults from a base class"?

Indeed this is also the plan. I think there is no way to get rid of all the function pointers. I have now partly implemented the filling out the struct thing now in the `functions` branch ([1](https://github.com/cwida/duckdb/blob/cfa94cf52255a479367bbf3cacdb221a08edf43c/src/function/aggregate_function/algebraic_functions.cpp), [2](https://github.com/cwida/duckdb/blob/cfa94cf52255a479367bbf3cacdb221a08edf43c/src/include/function/aggregate_function.hpp)), take a look and see how you think it looks.

**Comment by hawkfish** [comment]
[2](https://github.com/cwida/duckdb/blob/cfa94cf52255a479367bbf3cacdb221a08edf43c/src/include/function/aggregate_function.hpp) looks great!

My concern with [1](https://github.com/cwida/duckdb/blob/cfa94cf52255a479367bbf3cacdb221a08edf43c/src/function/aggregate_function/algebraic_functions.cpp) is that it scatters the definition of the aggregate function between files. I was wondering if a set of subclasses might be better, but that can wait.

Incidentally, there are two other aggregate state functions that might be needed (`merge` and `deallocate`) and I'm also thinking about how to resurrect the `AVG` optimisation, but that would just require more arguments. I've also never been happy about that silly `cast` Boolean - do you think this argument binding could get rid of that somehow?

**Comment by Mytherin** [comment]
Yes it should. The reason it was there was before we had auto-casting performing COUNT(a) on a string column would result in a cast of string to BIGINT. In the new system we would have COUNT(STRING) : BIGINT defined, which means the input will not be cast anymore for COUNT aggregates to begin with.

**Comment by Mytherin** [comment]
Implemented in c29057f91cd041912e8362312ef7d691dd00f747
